Anastasia Tikhonova w1 APIs

iHonu opened the pull request.

nixbix8 commented on 1-JavaScript/Week3/homework/ex1-giveCompliment.js:
```diff
@@ -18,13 +18,26 @@ Full description at: https://github.com/HackYourFuture/Homework/tree/main/1-Java
 -----------------------------------------------------------------------------*/
-function giveCompliment(/* TODO parameter(s) go here */) {
-  // TODO complete this function
+function giveCompliment(name) {
```
Hi @iHonu, these changes may have been committed by mistake, as they are your homework from the JavaScript Week3.
Could you update your PR so it only contains the code for the currently active homework, ie. 3-UsingAPIs/Week1?

nixbix8 commented on 3-UsingAPIs/Week1/homework/ex1-johnWho.js:
```diff
@@ -8,18 +8,19 @@ Rewrite this function, but replace the callback syntax with the Promise syntax:
+    setTimeout(() => {
+      if (!firstName) {
+        reject(new Error("You didn't pass in a first name!"));
+        return;
```
This is not really a mistake, but more for `those who want to know more` 😄  : The `return` statement after the `resolve` call is not necessary because it is unreachable code. Once a promise is resolved or rejected, the execution of the promise function is terminated and any code that comes after it will not be executed. Therefore, the return statement could be removed here.

nixbix8 commented on 3-UsingAPIs/Week1/homework/ex3-rollDie.js:
```diff
@@ -11,52 +11,47 @@ Full description at: https://github.com/HackYourFuture/Homework/tree/main/3-Usin
+      const value = Math.floor(Math.random() * 6) + 1;
+      console.log(`Die value is now: ${value}`);
+
+      // Use callback to notify that the die rolled off the table after 6 rolls
```
This comment should to be updated. 
Always try to make sure that the code comments reflect what is actually happening in the code. So when you update the code, make sure to update the comments around it as well. Stale comments might get quite misleading and confusing for someone looking at the code later.

nixbix8 commented on 3-UsingAPIs/Week1/homework/ex3-rollDie.js:
```diff
@@ -11,52 +11,47 @@ Full description at: https://github.com/HackYourFuture/Homework/tree/main/3-Usin
 }
 module.exports = rollDie;
+
+// There is no problem now because promise can return only one time
```
To clarify further, when using callbacks, if the die rolls off the table, the error callback is called repeatedly, once for each roll that exceeds the limit of 6 rolls. This happens because the `rollOnce()` function is called recursively using `setTimeout()` in a loop until all the rolls are done, and there is no way to stop the loop once the error callback is called.

However, in the solution with Promises, this issue does not happen  because, like you said, Promises can only be resolved or rejected once. So, even though the `rollOnce()` function is also called recursively, instead of calling a callback function, it either resolves the Promise with the final die value or rejects the Promise with an error message. This ensures that the Promise is resolved or rejected only once, regardless of how many times `rollOnce()` is called recursively.

iHonu commented:
Hello, thank you for your very quick feedback. You gave a detailed response, fixing the mistakes was easy, and the answer to the question clarified the confusion in that question. I fixed my mistakes and deleted the extra files

nixbix8 commented:
Good job Anastasia, really well done hw with some minor issues. Also, very quick in addressing and implementing the review points :)
